### ISSUE TITLE
Proposal to add grok parser operator

### ISSUE THREAD

**Issue by nityanandagohain** [open-telemetry/opentelemetry-collector-contrib#17184]
### Is your feature request related to a problem? Please describe.

Right now we have parser like json_parser, regex_parser which gets a lot of the work done. But using regex for parsing different kinds of logs increases the complexity and is difficult to maintain.

### Describe the solution you'd like

A grok parser operator can come to the rescue here. We can build on top of https://github.com/vjeantet/grok

### Additional context

If this is approved, I will be happy to pick this up and contribute.

**Comment by djaglowski** [comment]
I'm supportive of the general idea here. I think we need to pin down a few more details though.

First, are there any other grok libraries that we should be considering?

If we use the suggested library, my main concerns are related to the default patterns.

OTel uses a different naming style than the pattern "names" in the library. We'd want to alias these somehow so that the output is more OTel-like. In some cases, the keys would align with those defined in the OTel semantic conventions. 

It's not clear that users will want every naming pattern to be enabled by default. I imagine in some cases this would result in the operator doing "too much", clobbering other keys, etc. Because of this, I think we should look for ways to the user could easily specify the patterns they want to enable. For example, they could list the patterns they would like to enable: `patterns: [ip, ipv4, hostport]`, or groups of patterns: `pattern_groups: [networking]`.

**Comment by nityanandagohain** [comment]
We found a library that we can use https://github.com/vjeantet/grok , but since it was outdated and had minor issues we created a fork https://github.com/SigNoz/grok.


We maintain a fork of otel-collector where we have added support for this operator https://github.com/SigNoz/signoz-otel-collector/tree/main/pkg/parser/grok 

We use it similar to how we use regex_parser.

Though one of the problem we face right now is that we have to use the `router` operator which only supports regex.


For default patterns I think we will have to maintain it or allow the user to load some predefined patterns as well.
